### PR TITLE
Wooden floor burning + misc fire tweaks

### PR DESCRIPTION
**Pull request by xXxEMBERGHOSTSQUADDIExXx**
## What this does
wooden floors currently have an autoignition temperature set but the fire system isn't set up to actually burn turfs as fire fuel. this makes combustable turfs able to burn as fire fuel and also does a few more things:

- fixes burning wood floors not glowing properly
- fixes wooden floors not extinguishing properly
- makes flamethrowers light objects on fire
- makes increased oxygen in the air decrease autoignition temperature to a point
- adds a time-staggering effect to stuff autoigniting so it looks more organic
- lowers the wood floor fire_fuel from 10 to 5 to somewhat moderate the increased danger of burning the bar, library, courtroom etc.

https://github.com/vgstation-coders/vgstation13/assets/149450117/dd658af6-9cc2-4571-adb3-4c02e4b4a208

this does make fires in wooden rooms a good bit more dangerous, i leave that to your consideration and if you test it and think its too much we can maybe reduce their fire_fuel or change it another way.

note that carpet turfs arent actually combustible, the burned overlay they get is more of a decal thing. i didnt change anything with carpets here

## Why it's good
allow wooden floors to burn as you'd expect, more fire and chaos

## Changelog
:cl:
 * rscadd: Wooden floors can burn properly.
 * rscadd: Fire flashover looks hopefully more organic.
 * rscadd: Increased oxygen in the air makes things autoignite at lower temperatures, up to a point.
 * rscadd: Flamethrowers can ignite objects.
 * bugfix: Fixed burning wood floors not glowing properly.
 * bugfix: Fixed burning wood floors not extinguishing properly.
 * tweak: Halved fire fuel content of wooden floors.

### CODE REVIEW

**Comment by jwhitak** [comment]
No more fire. Please.

**Comment by Eneocho** [comment]
More fire, but reduce the amount of smoke produced (or make it transparent half the time). We have the pretty multicolour spectra for burning stuff and most of the thing is obscured by smoke.
I want to see the station glow white in hellfire.

**Comment by GreatBigFailure** [comment]
What's leftover after they burn away? I hope it's the metal under-tile flooring and not space.
If it doesn't space the room then good change, fires are fun. 

>More fire, but reduce the amount of smoke produced

I agree with Eneocho though, reduce the smoke or make it lean toward a smaller amount if there's random variance. The smoke can be a nice touch but does often obscure too much.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
>What's leftover after they burn away

The "burned wooden floor" tile, which is still the wooden floor tile but not further flammable.
image

**Comment by GreatBigFailure** [comment]
>The "burned wooden floor" tile, which is still the wooden floor tile but not further flammable.

Oh, perfect then. Great change.

**Comment by brndd** [comment]
I like the idea of wood floors being combustible, but they shouldn't catch fire as easily as other flammable things do (and they look very combustible in the video).

If you've ever tried to set fire to a big log without kindling in real life, you'll know that a solid block of wood doesn't catch fire easily. Of course if it reaches a high enough temperature (modeled here by the autoignition temperature) it'll catch fire, but wood has a fair bit of thermal mass and isn't a great conductor of heat, so this does not happen quickly. I don't think our system models either of those factors, so wood floors instantly combusting when the environment they're in reaches AUTOIGNITION_WOOD is less than realistic.

Basically I'd like fires to work something like real housefires. Carpets, curtains, papers, plastics, molotov cocktails etc. start the fire, and if it gets big enough off of those the actual structure (in this case, wood floors) may start burning as well -- but it's very difficult for a candle dropped on bare floorboards to actually start a fire.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
good point. 
In that video I opened a canned heat

I think for our purposes if the air atmos contents temperature reaches the autoignition point it should combust, but with hotspot expose (matches, dropped cigarettes etc) there should maybe be be some considerations like you mentioned

theres already some functionality like that: a match doesn't always instantly ignite a wood floor because there is a per exposure chance based on the exposed volume,

i left a lit cigarette on the floor of the bar and it stayed there about a minute before it caught the floor on fire

But there should probably be some additional factor like you said. Maybe something that influences the spread speed from different materials

I'll think about how to do that

**Comment by content-to-remain-anonymous** [comment]
the obvious solution to this problem is adding heatflow and combustion simulation to sparks and cigarettes
a real lit cigarette won't ignite if you throw it into a pile of diesel, a hardwood floortile combusting is highly silly
currently i believe there's also no downside to replacing the distro with 100% oxygen where realistically it would turn your cigarettes into bombs

and add slag!

**Comment by Eneocho** [comment]
I vaguely remember that full oxygen rooms can last longer while on fire.

**Comment by brndd** [comment]
> I vaguely remember that full oxygen rooms can last longer while on fire.

This is only logical because they have more oxygen to feed the fire. But I don't think oxygen makes it any easier to start fires currently, and it really should because it would be both realistic and hilarious.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
so maybe the oxygen content should factor into it.. or there is also a volatility var but basically every atom has the same volatility. that could be factored in too.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
draft to consider these:

- [x] certain edge case
- [x] air oxygen content
- [ ] revisit spark temperature in light of the above
- [ ] other situations where it makes sense to shift the autoignition temp
- [ ] kuzdu
- [ ] object volatility
- [ ] smoke frequency

**Comment by GreatBigFailure** [comment]
Not for this PR of course, but since you're doing all this fire stuff already: Any chance of making fire interact with kudzu and other vines? It's always disappointing when a desperation welder fuel fire doesn't actually work. You'd think it would at least damage them.

**Comment by Eneocho** [comment]
But it does affect them. If the room is too hot the vines stop growing.
I've stopped a flood with me trusty ZAS Gun more than once.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
If that wasnt the case I could see there being a Conways game of life type thing with vines spreading and fire burning them away perpetually

**Comment by Eneocho** [comment]
> If that wasnt the case I could see there being a Conways game of life type thing with vines spreading and fire burning them away perpetually

That can still happen if you pump the heat resistance for the plant high enough, and temperature manipulation (to avoid the room getting to 50000K or something)

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
tangentially, here and there in the atmos code I sometimes see references to the idea of heat transferring through walls
but was that ever active and it was disabled or broken at some point?

**Comment by Eneocho** [comment]
I'd like to preemptively state that I do not want heat transferring through walls.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
The smoke thing I was thinking that hotter fires could make less smoke as they more fully combust the materials or something like that. 
But then it also makes me want for there to be partial combustion byproducts as gases and soot/smoke as a quasi-gas maybe

**Comment by content-to-remain-anonymous** [comment]
flue gas? syngas? charcoal piles? petrochem gas reactions?!
sign me up

**Comment by Eneocho** [comment]
Can't wait for the NT petrochem spill of 2658 that contaminates half of explorable space and has the odd side effect of making pengpeng reproduce like a rabbit (only 2000s kids will get this reference).

**Comment by Eneocho** [comment]
> The smoke thing I was thinking that hotter fires could make less smoke as they more fully combust the materials or something like that. 
> But then it also makes me want for there to be partial combustion byproducts as gases and soot/smoke as a quasi-gas maybe
> 

Get out of my FUCKING HEAD REEEEEE. I was gonna suggest that hotter fires make less smoke due to more complete combustion and smoke mechanics on oxygen rich/deficient rooms.
An incomplete burn could make CO instead of CO2, indirectly poisoning the room, with the ghetto cleanup solution being flooding the room with O2 and burning so the CO becomes CO2.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
i based on the oxygen stuff on this:
https://sci-hub.se/https://doi.org/10.1016/j.jlp.2019.103971
image

at double standard station oxygen, autoignition temperature is 30% lower. that feels reasonable and gets paper and wood to about 360 and 400K. beyond that the effect is capped

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
About the smoke stuff I very much like the idea of fires spilling massive amounts of gunk into the confined space that is the station air and wreaking havoc on the atmos infrastructure

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
I think the other changes listed above merit their own dedicated prs, some possibly after object material composition is standardized, along with their own voting so I'll open this

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
>make stuff burn faster with more oxygen

Sounds nice and I can look into this. A little tricky maybe without doing a deeper rework.

>raise AIG temp if there's very little oxygen

currently happens with the way I did the lowering

>slow/stop burning if there's no oxygen

Currently happens generally except for some potential edge cases, i think i know what to change to fix those (fixed)

>let players see how close burning items are to completely burning up(turn sprite blacker as fire_fuel approaches 0?)

Maybe but idk if I'll do that here. Could a burned item be restored?


>turn burnt objects into renamed slag (and only make ash when at REALLY high temps/burning organics)

Seems good but idk if I'll do that here
Also I wanna make a note of that 'burned box doesn't automatically delete its contents upon ashification' thing
   
>and you should probably switch to partial oxygen pressure - real fires burn better at higher pressures and don't burn in vacuum

this is already essentially how it works with the current molar density, as that factors in  volume and is directly related to partial oxygen pressure. I think the rest of the desired behavior would be implemented as a matter of course if the 'burn faster with more oxygen' you mentioned is added

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
@Exxion how would I typecast the return value of return_air()?

File: code/game/atoms.dm [hunk 1]
```diff
@@ -198,7 +198,7 @@ var/global/list/ghdel_profiling = list()
 /atom/proc/remove_air(amount)
 	return null
 
-/atom/proc/return_air()
+/atom/proc/datum/gas_mixture/return_air()
```
Comment (Exxion) [inline code/game/atoms.dm]:
```suggestion
/atom/proc/return_air() as /datum/gas_mixture
```

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
thanks!

**Comment by Exxion** [comment]
Oh, I guess the linter still doesn't actually know how to deal with `as` return types. Wonder when they'll hook that up.

**Comment by xXxEMBERGHOSTSQUADDIExXx** [comment]
the linter is good i guess but it does get in the way sometimes

it kinda blocks us from having fun with the ducky nature of byond

**Comment by Kurfursten** [comment]
This now has conflicts and it seems very unlikely that Hina will return to fix them. You can let us know if you want this reopened.
